### PR TITLE
Add add_pair feature

### PR DESCRIPTION
**Pull request by gohyongjing**
Added addPair command
* Format: `add_pair elderly/ELDERLY_NAME volunteer/VOLUNTEER_NAME`

Updated user interface to split screen between person list and pair list
photo_6285080649786307797_w (1)

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/AddPairCommand.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import seedu.address.logic.commands.exceptions.CommandException;
+import seedu.address.model.Model;
+import seedu.address.model.pair.Pair;
+
+/**
+ * Adds a person to the address book.
+ */
+public class AddPairCommand extends Command {
+
+    public static final String COMMAND_WORD = "add_pair";
```
Comment (Zxun2) [inline src/main/java/seedu/address/logic/commands/AddPairCommand.java:16]:
```suggestion
    public static final String COMMAND_WORD = "add:pair";
```
on second thought, this syntax is more soothing to the eye 😄 what do the rest think?

File: src/main/java/seedu/address/logic/parser/CliSyntax.java [hunk 2]
```diff
@@ -11,5 +11,7 @@ public class CliSyntax {
     public static final Prefix PREFIX_EMAIL = new Prefix("e/");
     public static final Prefix PREFIX_ADDRESS = new Prefix("a/");
     public static final Prefix PREFIX_TAG = new Prefix("t/");
+    public static final Prefix PREFIX_ELDERLY = new Prefix("elderly/");
+    public static final Prefix PREFIX_VOLUNTEER = new Prefix("volunteer/");
```
Comment (Zxun2) [inline src/main/java/seedu/address/logic/parser/CliSyntax.java:15]:
```suggestion
    public static final Prefix PREFIX_ELDERLY = new Prefix("ent:e/");
    public static final Prefix PREFIX_VOLUNTEER = new Prefix("ent:v/");
```
we can do this, so there will be no ambiguity between similar alphabets. i think this is better too, as it provides greater clarity on the flags we're specifying. 

perhaps we should create an issue to reorganize all the existing commands such that:
1. flags follow `flags:f/`
2. arguments follow `args:a/`
3. entities follow `ent:e/`

what do you think? others feel free to chime in

File: src/main/java/seedu/address/logic/parser/AddPairCommandParser.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+                || !argMultimap.getPreamble().isEmpty()) {
+            throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT, AddPairCommand.MESSAGE_USAGE));
+        }
+
+        Name elderly = ParserUtil.parseName(argMultimap.getValue(PREFIX_ELDERLY).get());
+        Name volunteer = ParserUtil.parseName(argMultimap.getValue(PREFIX_VOLUNTEER).get());
+        Phone phone = ParserUtil.parsePhone("88888888");
+        Email email = ParserUtil.parseEmail("dummy@email.com");
+        Address address = ParserUtil.parseAddress("dummy address");
+        Set<Tag> tagList = ParserUtil.parseTags(new HashSet<>());
```
Comment (Zxun2) [inline src/main/java/seedu/address/logic/parser/AddPairCommandParser.java:45]: is this needed? or for testing? 
```suggestion
        Phone phone = ParserUtil.parsePhone("88888888");
        Email email = ParserUtil.parseEmail("dummy@email.com");
        Address address = ParserUtil.parseAddress("dummy address");
        Set<Tag> tagList = ParserUtil.parseTags(argMultimap.getAllValues(PREFIX_TAG));
```

File: src/main/java/seedu/address/logic/parser/AddPairCommandParser.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        return new AddPairCommand(pair);
+    }
+
+    /**
+     * Returns true if none of the prefixes contains empty {@code Optional} values in the given
+     * {@code ArgumentMultimap}.
+     */
+    private static boolean arePrefixesPresent(ArgumentMultimap argumentMultimap, Prefix... prefixes) {
+        return Stream.of(prefixes).allMatch(prefix -> argumentMultimap.getValue(prefix).isPresent());
+    }
```
Comment (Zxun2) [inline src/main/java/seedu/address/logic/parser/AddPairCommandParser.java:59]: seems like a duplicate method. better to extract it out to utils and name it `PrefixUtils` or something

File: src/main/java/seedu/address/model/AddressBook.java [hunk 5]
```diff
@@ -106,15 +146,24 @@ public ObservableList<Person> getPersonList() {
...
+        return pairs.asUnmodifiableObservableList();
+    }
+
     @Override
     public boolean equals(Object other) {
         return other == this // short circuit if same object
                 || (other instanceof AddressBook // instanceof handles nulls
                 && persons.equals(((AddressBook) other).persons));
+        // todo (yong jing)
+                //&& pairs.equals(((AddressBook) other).pairs));
```
Comment (Zxun2) [inline src/main/java/seedu/address/model/AddressBook.java:160]: can just write `// TODO: task description`and remove the implementation

Comment (Zxun2) [reply src/main/java/seedu/address/model/AddressBook.java:160]: same for the rest

File: src/main/java/seedu/address/model/pair/Pair.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+import java.util.Objects;
+
+import seedu.address.model.person.Person;
+
+/**
+ * Represents a Pair of elderly and volunteer in the address book.
+ * Guarantees: details are present and not null, field values are validated, immutable.
+ */
+public class Pair {
```
Comment (Zxun2) [inline src/main/java/seedu/address/model/pair/Pair.java:13]: can you create an attribute for the `UUID`as well? maybe also extends from an interface, `Identifiable` that contains the basic getters and setters.

File: src/main/java/seedu/address/model/pair/UniquePairList.java [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+ *
+ * Supports a minimal set of list operations.
+ *
+ * @see Pair#isSamePair(Pair)
+ */
+public class UniquePairList implements Iterable<Pair> {
+
+    private final ObservableList<Pair> internalList = FXCollections.observableArrayList();
+    private final ObservableList<Pair> internalUnmodifiableList =
+            FXCollections.unmodifiableObservableList(internalList);
```
Comment (Zxun2) [inline src/main/java/seedu/address/model/pair/UniquePairList.java:29]: honestly, i don't think we should resort to using the current list implementation. if possible, can you make use of hashmaps instead? if not, we can KIV this, and create a corresponding issue for the future

File: src/main/java/seedu/address/model/pair/exceptions/DuplicatePairException.java [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+
+/**
+ * Signals that the operation will result in duplicate Pairs (Pairs are considered duplicates if they have the same
+ * identity).
+ */
+public class DuplicatePairException extends RuntimeException {
+    public DuplicatePairException() {
+        super("Operation would result in duplicate pairs");
+    }
+}
```
Comment (Zxun2) [inline src/main/java/seedu/address/model/pair/exceptions/DuplicatePairException.java:11]: can you abstract this to `DuplicateEntityException` that takes in the `errMsg` and `Class<?>`? the same logic for checking duplicates seems to be duplicated (no puns intended 😄) at various parts of the codebase. 

what do you think of adding a public static method `handleException(boolean condition)` to replace it?

File: src/test/java/seedu/address/testutil/TypicalPairs.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+    /**
+     * Returns an {@code AddressBook} with all the typical pairs.
+     */
+    public static AddressBook getTypicalAddressBook() {
+        AddressBook ab = new AddressBook();
+        for (Pair pair : getTypicalPairs()) {
+            ab.addPair(pair);
+        }
+        return ab;
+    }
```
Comment (Zxun2) [inline src/test/java/seedu/address/testutil/TypicalPairs.java:38]: change it to `FriendlyLink`

**Review by Zxun2 — CHANGES_REQUESTED** [review]
well done! logic is generally good. i left some comments below, waiting to hear your thoughts!

File: src/main/java/seedu/address/logic/parser/AddPairCommandParser.java (hunk 3 above)
Comment (gohyongjing) [reply src/main/java/seedu/address/logic/parser/AddPairCommandParser.java:45]: This is a placeholder, will replace it with a call to getElderly or something once the relevant classes are pulled

File: src/main/java/seedu/address/model/pair/Pair.java (hunk 6 above)
Comment (gohyongjing) [reply src/main/java/seedu/address/model/pair/Pair.java:13]: I believe this is already done by heyi and shaun, as the NRIC class

File: src/main/java/seedu/address/model/pair/UniquePairList.java (hunk 7 above)
Comment (gohyongjing) [reply src/main/java/seedu/address/model/pair/UniquePairList.java:29]: We will keep this in view

File: src/main/java/seedu/address/logic/commands/AddPairCommand.java (hunk 1 above)
Comment (gohyongjing) [reply src/main/java/seedu/address/logic/commands/AddPairCommand.java:16]: Lets discuss this in the next meeting

File: src/main/java/seedu/address/logic/parser/CliSyntax.java (hunk 2 above)
Comment (gohyongjing) [reply src/main/java/seedu/address/logic/parser/CliSyntax.java:15]: Lets discuss this in the next meeting

File: src/main/java/seedu/address/model/pair/exceptions/DuplicatePairException.java (hunk 8 above)
Comment (gohyongjing) [reply src/main/java/seedu/address/model/pair/exceptions/DuplicatePairException.java:11]: Using DuplicateEntityException<T> will mean will will need a way to retrieve the plural word of each noun, like "pairs" and "persons", will keep it in view

File: src/main/java/seedu/address/logic/parser/AddPairCommandParser.java (hunk 4 above)
Comment (gohyongjing) [reply src/main/java/seedu/address/logic/parser/AddPairCommandParser.java]: Fixed in commit 67672a75856ccd60cf068640ae88cefda7f8cc2d

File: src/main/java/seedu/address/model/AddressBook.java (hunk 5 above)
Comment (gohyongjing) [reply src/main/java/seedu/address/model/AddressBook.java]: Fixed in commit 4bc2292

File: src/test/java/seedu/address/testutil/TypicalPairs.java (hunk 9 above)
Comment (gohyongjing) [reply src/test/java/seedu/address/testutil/TypicalPairs.java]: Fixed in commit 4bc2292

File: src/main/java/seedu/address/model/pair/Pair.java (hunk 6 above)
Comment (Zxun2) [reply src/main/java/seedu/address/model/pair/Pair.java:13]: what's should be the current syntax for deleting a pair? are we specifying the pair or the two entities?

File: src/main/java/seedu/address/logic/parser/CliSyntax.java (hunk 2 above)
Comment (Zxun2) [reply src/main/java/seedu/address/logic/parser/CliSyntax.java:15]: for now can you update this to be `enr/` and `vnr/` respectively?

File: src/main/java/seedu/address/model/pair/Pair.java (hunk 6 above)
Comment (gohyongjing) [reply src/main/java/seedu/address/model/pair/Pair.java:13]: If we use entities we would need to specify 2 different ids. If we specify a pair, we can use the id of the pair. Id of pair seems easier to use and quicker to type

File: src/main/java/seedu/address/logic/parser/CliSyntax.java (hunk 2 above)
Comment (gohyongjing) [reply src/main/java/seedu/address/logic/parser/CliSyntax.java]: Fixed in commit 954655f56d0a76c3da628c0304088eb3b4dbc54a

File: src/main/java/seedu/address/model/pair/Pair.java (hunk 6 above)
Comment (gohyongjing) [reply src/main/java/seedu/address/model/pair/Pair.java:13]: Discussed in meeting 4

**Review by Zxun2 — APPROVED** [review]
Looking good 👍🏻 if the rest is okay with this, we can merge it in!

**Comment by Zxun2** [comment]
@gohyongjing Should we merge this in?

**Comment by gohyongjing** [comment]
Done! Thanks guys

**Comment by gohyongjing** [comment]
Done! Thanks everyone

**Pull request merged**